src/storage/src/render/upsert.rs

```diff
@@ -8,1045 +8,274 @@
-        >,
-    >,
-) {
-    let mut session = output.session(cap);
-    removed_times.push(time.clone());
-    for (key, data) in std::mem::take(map) {
-        // decode key and value, and apply predicates/projections to they combined key/value
-        if let Some(decoded_key) = key {
-            let (decoded_key, decoded_value): (_, Result<_, DataflowError>) =
-                match (decoded_key, data.value) {
-                    // Make sure that the value is also an Err if they key is an
-                    // Err. Otherwise, any decoding logic that would try and
-                    // work on this can get flustered.
-                    (Err(upsert_error), Some(_)) => (
-                        Err(upsert_error.clone()),
-                        Err(EnvelopeError::Upsert(upsert_error).into()),
-                    ),
-                    (Err(upsert_error), None) => (Err(upsert_error), Ok(None)),
-                    (Ok(decoded_key), None) => (Ok(decoded_key), Ok(None)),
-                    (Ok(decoded_key), Some(value)) => {
-                        let decoded_value = value
-                            .map_err(|err| EnvelopeError::Upsert(err).into())
-                            .and_then(|row| {
-                                build_datum_vec_for_evaluation(
-                                    dv,
-                                    &upsert_envelope.style,
-                                    &row,
-                                    &decoded_key,
-                                )
-                                .map_or(Ok(None), |mut datums| {
-                                    datums.extend(data.metadata.iter());
-                                    evaluate(&datums, predicates, position_or, row_packer)
-                                        .map_err(DataflowError::from)
-                                })
-                            });
-                        (Ok(decoded_key), decoded_value)
+        // Now can can resume consuming the collection
+        let mut pending_batches = vec![];
+        let mut output_updates = vec![];
+        let mut input_upper = Antichain::from_elem(Timestamp::minimum());
+        while let Some(event) = input.next_mut().await {
+            match event {
+                AsyncEvent::Data(_cap, data) => {
+                    if PartialOrder::less_equal(&input_upper, &resume_upper) {
+                        data.retain(|(_, ts, _)| resume_upper.less_equal(ts));
                     }
-                };
 
-            // Turns Ok(None) into None, and others into Some(OK) and Some(Err).
-            // We store errors as well as non-None values, so that they can be
-            // retracted if new rows show up for the same key.
-            let new_value = decoded_value.transpose();
+                    data.sort_unstable_by(|((_, a_order), a_ts, _), ((_, b_order), b_ts, _)| {
+                        a_ts.cmp(b_ts).then_with(|| a_order.cmp(b_order))
+                    });
```

Comment:
I can describe what it does and maybe you have an idea for a better name. It's true that it's currently the kakfa offset but as far as this function is concerned it is a generic parameter that specifies the order of commands when they happen on the same timestamp and after Mouli's work it might be something else than the offset so I don't think it should be called `offset`.